### ISSUE TITLE
[DataPipe] Ensure all DataPipes Meet Testing Requirements

### ISSUE THREAD

**Issue by NivekT** [pytorch/data#106]
## 🚀 Feature

We have many tests for existing DataPipes (both in PyTorch Core and TorchData). However, over time, they have become less organized. Moreover, as the [testing requirements](https://github.com/pytorch/pytorch/wiki/DataPipes-testing-requirements) expand, older DataPipes may not have tests to cover the newly added requirements.

This issue aims to track the status of tests for all DataPipes.

## Motivation

We want to ensure test coverage for all DataPipe is complete to reduce bugs and unexpected behavior.

## Alternative

We also should create some testing templates for `IterDataPipe` and `MapDataPipe` that can be widely applied.

## `IterDataPipe` Tracker

X - Done
NA - Not Applicable
Blank - Unclear

| Name | Module | Functional Test | Reset | `__len__` | Serializable | Pickable | Graph
| ------------- | ------------- | ------------- | ------------- | ------------- | ------------- | ------------- | ------------- |
| Batcher | Core |  
| BucketBatcher | Core |
| Collator | Core |
| Concater | Core |
| Demultiplexer | Core |
| FileLister | Core |
| FileLoader | Core |
| Filter | Core |
| Forker | Core |
| Grouper | Core |
| IterableWrapper | Core |
| Mapper | Core |
| Multiplexer | Core |
| RoutedDecoder | Core |
| Sampler | Core |
| Shuffler | Core |
| StreamReader | Core |
| UnBatcher | Core |
| Zipper | Core |
| CSVDictParser | Data | 
| CSVParser | Data | 
| Cycler | Data | 
| Enumerator | Data | 
| Extractor | Data | 
| GDriveReader | Data | 
| HashChecker | Data | 
| Header | Data | 
| HttpReader | Data | 
| InMemoryCacheHolder | Data | 
| IndexAdder | Data | 
| IoPathFileLister | Data | 
| IoPathFileLoader | Data | 
| IterKeyZipper | Data | 
| JsonParser | Data | 
| LineReader | Data | 
| MapKeyZipper | Data | 
| OnDiskCacheHolder | Data | 
| OnlineReader | Data | 
| ParagraphAggregator | Data | 
| Rows2Columnar | Data | 
| SampleMultiplexer | Data | 
| Saver | Data | 
| TarArchiveReader | Data | 
| XzFileReader | Data | 
| ZipArchiveReader | Data | 

## `MapDataPipe` Tracker

To be created when we have more `MapDataPipes`.

**Comment by ejguan** [comment]
This is awesome. One nit note: serializable should be same as picklable IMO.

**Comment by VitalyFedyunin** [comment]
We can require each DataPipe to introduce a simple usage example graph for this purpose.

**Comment by ejguan** [comment]
When we have time, we might need to go over our DataPipes again to identify any missing test.

Besides, for future reference, we might need to improve our testing framework to something similar to `OpInfo` in PyTorch Core to run the testing coverage automatically without we go over each test by ourselves.

**Comment by NivekT** [comment]
> When we have time, we might need to go over our DataPipes again to identify any missing test since there are a few DataPipe implemented recently.
> 
> Besides, for future reference, we might need to improve our testing framework to something similar to `OpInfo` in PyTorch Core to run the testing coverage automatically without we go over each test by ourselves.

Agreed that the `OpInfo`-like way is probably the best. I think our inputs and necessary setup for each test is a bit all over the place. Having tests split between two repos doesn't help either.
